gateways/stripe/stripe_vars.php

```diff
@@ -2,17 +2,58 @@
 
 function espresso_display_stripe($payment_data) {
 	extract($payment_data);
-	global $org_options;
+	global $org_options, $wpdb;
 	$stripe_settings = get_option('event_espresso_stripe_settings');
 	if ($stripe_settings['force_ssl_return']) {
-			$home = str_replace('http://', 'https://', home_url());
-		} else {
-			$home = home_url();
-		}
+		$home = str_replace('http://', 'https://', home_url());
+	} else {
+		$home = home_url();
+	}
+	
+	$SQL = "SELECT start_date FROM " . EVENTS_ATTENDEE_TABLE . " WHERE registration_id=%s ORDER BY id LIMIT 1";
+	$payment_data['start_date'] = $wpdb->get_var( $wpdb->prepare( $SQL, $payment_data['registration_id']));
 
-	wp_register_script( 'stripe', EVENT_ESPRESSO_PLUGINFULLURL . 'gateways/stripe/stripe.js', array( 'jquery.validate.js' ), '1.0', TRUE );
-	wp_enqueue_script( 'stripe' );	
+	$stripe_description = sprintf(
+		/* translators: 1: event name, 2: event date, 3: attendee first name, 4: attendee last name */
+		__('%1$s [%2$s] >> %3$s %4$s', 'event_espresso'),
+		$payment_data["event_name"],
+		date('m-d-Y', strtotime($payment_data['start_date'])),
+		$payment_data["fname"],
+		$payment_data["lname"]
+	);
+	
+	//Check for an alternate Stripe settings
+	if ( !empty($payment_data['event_meta']['stripe_publishable_key'] ) ) {
+		//Alternate Stripe settings
+		$stripe_settings['stripe_publishable_key'] = $payment_data['event_meta']['stripe_publishable_key'];
+	}
 
+	wp_enqueue_script('stripe_payment_js', 'https://js.stripe.com/v3/', array(), false, true);
+	wp_enqueue_script('espresso_stripe', EVENT_ESPRESSO_PLUGINFULLURL . 'gateways/stripe/espresso_stripe.js', array('stripe_payment_js'), false, true);
+	wp_enqueue_style('espresso_stripe_css', EVENT_ESPRESSO_PLUGINFULLURL . 'gateways/stripe/espresso_stripe.css');
+
+	//Include the Stripe API
+	require_once (dirname(__FILE__).'/stripe-php-6.43.1/init.php');
+	// Setup a Stripe object using the secret key.
+	\Stripe\Stripe::setApiKey( $stripe_settings['stripe_secret_key']);
+
+	$payment_intent_id = isset($_REQUEST['espresso_stripe_payment_intent_id']) ? $_REQUEST['espresso_stripe_payment_intent_id'] : '';
+	if( $payment_intent_id ) { 
+		$intent = \Stripe\PaymentIntent::retrieve(
+            $payment_intent_id
+        );
+	} else {
+		$intent = \Stripe\PaymentIntent::create([
+	    	'amount' => str_replace( array(',', '.'), '', number_format( $event_cost, espresso_get_stripe_decimal_places($stripe_settings['stripe_currency_symbol'])) ),
+	     	'currency' => $stripe_settings['stripe_currency_symbol'],
+		], [
+			'idempotency_key' => $registration_id,
```

Comment:
> If the registration ID is used to identify a unique payment intent, what happens if a payment fails and is retried?

Each payment creates a charge on the payment intent (PI) object, so it doesn't need a new PI each time, just the ID of the PI to assign the charge to (which obviously doesn't change).

> Won't Stripe see the duplicate idempotency_key and assume it's a repeat of the failed payment?

No, because the idempotency_key is only being used to create a payment intent itself, the charges on that PI are created/handled by `handleCardPayment` and don't use it the above. So even with the idempotency_key set the only thing that happens is you get multiple charges on the PI, like so - https://monosnap.com/file/GgUChlXkUZymx0G6ArPupPNVw92usG

That's just from me forcing 3DS failures and then trying again (without refreshing the page and using the same card details), then refreshing the page and adding the same details, this time completing the 3DS prompt.

The above just stops the user getting to the payment options and being able to refresh the page X times ending up with X number of PI's on the Stripe account that then aren't used.

Reply:
Hmmm I think I kinda follow. It sounds like I could use that for EE4's integration too then.